File: covalent_ui/webapp/src/components/Result.js
/**
 * Copyright 2021 Agnostiq Inc.
 *
 * This file is part of Covalent.
 *
 * Licensed under the GNU Affero General Public License 3.0 (the "License").
 * A copy of the License may be obtained with this software package or at
 *
 *      https://www.gnu.org/licenses/agpl-3.0.en.html
 *
 * Use of this file is prohibited except in compliance with the License. Any
 * modifications or derivative works of this file must retain this copyright
 * notice, and modified files must contain a notice indicating that they have
 * been altered from the originals.
 *
 * Covalent is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
 *
 * Relief from the License may be granted by purchasing a commercial license.
 */

import _ from 'lodash'
import { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useParams } from 'react-router-dom'
import ReactFlow, { Controls, MiniMap } from 'react-flow-renderer'
import { Helmet } from 'react-helmet-async'
import {
  Box,
  Divider,
  ToggleButton,
  ToggleButtonGroup,
  Toolbar,
  Paper,
  Breadcrumbs,
  Link,
  Typography,
  IconButton,
} from '@mui/material'
import {
  ArrowBack,
  ArrowDownward,
  ArrowForward,
  ArrowUpward,
  Map as MapIcon,
  Refresh,
} from '@mui/icons-material'

import NodeInfo from './NodeInfo'
import ElectronNode from './graph/ElectronNode'
import ParameterNode from './graph/ParameterNode'
import layout from './graph/Layout'
import CopyButton from './CopyButton'
import { fetchResult } from '../redux/resultsSlice'

const Result = () => {
  const { dispatchId } = useParams()
  const result = useSelector((state) => state.results.cache[dispatchId])

  const [elements, setElements] = useState()
  const [direction, setDirection] = useState('TB')
  const [showMinimap, setShowMinimap] = useState(false)
  const [showParams, setShowParams] = useState(true)
  const [selectedNodeId, setSelectedNodeId] = useState(null)
  const [flowInstance, setFlowInstance] = useState()

  const dispatch = useDispatch()
  const handleRefresh = () => {
    dispatch(
      fetchResult({
        dispatchId: result.dispatch_id,
        resultsDir: result.results_dir,
      })
    )
  }

  useEffect(() => {
    setElements(layout(result.graph, direction, showParams))
  }, [result, direction, showParams])

  useEffect(() => {
    setTimeout(() => {
      if (flowInstance) {
        flowInstance.fitView()
      }
    })
  }, [elements, flowInstance])

  // handle resizing
  useEffect(() => {
    const resizeHandler = _.debounce(() => {
      flowInstance && flowInstance.fitView()
    }, 150)

    window.addEventListener('resize', resizeHandler)
    return () => {
      window.removeEventListener('resize', resizeHandler)
    }
  })

  if (!result) {
    return <div>Not found</div>
  }

  return (
    <>
      <Helmet title={dispatchId} />

      <Breadcrumbs sx={{ mb: 1 }}>
        <Link underline="hover" color="inherit" href="/">
          Results
        </Link>
        <Typography color="text.primary" fontWeight="500">
          {dispatchId}
        </Typography>
      </Breadcrumbs>

      <Box
        sx={{
          color: 'text.secondary',
          fontSize: 'body2.fontSize',
          fontFamily: 'monospace',
        }}
      >
        {result.results_dir}
        <CopyButton
          content={result.results_dir}
          size="small"
          title="Copy results directory"
        />
      </Box>

      <Toolbar>
        <ToggleButtonGroup
          value={direction}
          onChange={(event, newDirection) => {
            if (newDirection) {
              setDirection(newDirection)
            }
          }}
          exclusive
          size="small"
          sx={{ marginLeft: 'auto' }}
        >
          <ToggleButton value="TB">
            <ArrowDownward />
          </ToggleButton>
          <ToggleButton value="BT">
            <ArrowUpward />
          </ToggleButton>
          <ToggleButton value="LR">
            <ArrowForward />
          </ToggleButton>
          <ToggleButton value="RL">
            <ArrowBack />
          </ToggleButton>
        </ToggleButtonGroup>

        <Divider orientation="vertical" sx={{ m: 2 }} flexItem />

        <ToggleButton
          value="checked"
          selected={showMinimap}
          onChange={() => setShowMinimap(!showMinimap)}
          size="small"
        >
          <MapIcon />
        </ToggleButton>

        <Divider orientation="vertical" sx={{ m: 2 }} flexItem />

        <ToggleButton
          value="checked"
          selected={showParams}
          onChange={() => setShowParams(!showParams)}
          size="small"
        >
          Parameters
        </ToggleButton>

        <Divider orientation="vertical" sx={{ m: 2 }} flexItem />

        <IconButton onClick={handleRefresh}>
          <Refresh />
        </IconButton>
      </Toolbar>

      <Paper sx={{ p: 2, height: '500px' }}>
        <ReactFlow
          nodeTypes={{ electron: ElectronNode, parameter: ParameterNode }}
          nodesDraggable={false}
          nodesConnectable={false}
          onLoad={(instance) => {
            setFlowInstance(instance)
          }}
          elements={elements}
          onSelectionChange={(elements) => {
            // assume no multi-selection for now
            setSelectedNodeId(_.get(elements, ['0', 'id'], null))
          }}
        >
          <Controls />
          {showMinimap && <MiniMap />}
        </ReactFlow>
      </Paper>

      <NodeInfo dispatchId={dispatchId} nodeId={selectedNodeId} />
    </>
  )
}

export default Result

File: covalent_dispatcher/_core/result.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the GNU Affero General Public License 3.0 (the "License").
# A copy of the License may be obtained with this software package or at
#
#      https://www.gnu.org/licenses/agpl-3.0.en.html
#
# Use of this file is prohibited except in compliance with the License. Any
# modifications or derivative works of this file must retain this copyright
# notice, and modified files must contain a notice indicating that they have
# been altered from the originals.
#
# Covalent is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
#
# Relief from the License may be granted by purchasing a commercial license.

"""
Defines the core functionality of the result service
"""

import uuid

from covalent._results_manager import Result
from covalent._shared_files import logger
from covalent._workflow.lattice import Lattice
from covalent_ui import result_webhook

from .._db import update

app_log = logger.app_log
log_stack_info = logger.log_stack_info


def generate_node_result(
    node_id,
    start_time=None,
    end_time=None,
    status=None,
    output=None,
    error=None,
    stdout=None,
    stderr=None,
    sublattice_result=None,
):

    return {
        "node_id": node_id,
        "start_time": start_time,
        "end_time": end_time,
        "status": status,
        "output": output,
        "error": error,
        "stdout": stdout,
        "stderr": stderr,
        "sublattice_result": sublattice_result,
    }


# Domain: result
async def _update_node_result(result_object, node_result, pending_deps, status_queue):
    app_log.warning("Updating node result (run_planned_workflow).")
    update._node(result_object, **node_result)
    await result_webhook.send_update(result_object)
    node_id = node_result["node_id"]
    node_status = node_result["status"]
    await status_queue.put((node_id, node_status))


# Domain: result
def initialize_result_object(
    json_lattice: str, parent_result_object: Result = None, parent_electron_id: int = None
) -> Result:
    """Convenience function for constructing a result object from a json-serialized lattice.

    Args:
        json_lattice: a JSON-serialized lattice
        parent_result_object: the parent result object if json_lattice is a sublattice
        parent_electron_id: the DB id of the parent electron (for sublattices)

    Returns:
        Result: result object
    """

    dispatch_id = get_unique_id()
    lattice = Lattice.deserialize_from_json(json_lattice)
    result_object = Result(lattice, lattice.metadata["results_dir"], dispatch_id)
    if parent_result_object:
        result_object._root_dispatch_id = parent_result_object._root_dispatch_id

    result_object._initialize_nodes()
    app_log.debug("2: Constructed result object and initialized nodes.")

    update.persist(result_object, electron_id=parent_electron_id)
    app_log.debug("Result object persisted.")

    return result_object


# Domain: result
def get_unique_id() -> str:
    """
    Get a unique ID.

    Args:
        None

    Returns:
        str: Unique ID
    """

    return str(uuid.uuid4())
